master/buildbot/scripts/devproxy.py

```diff
@@ -0,0 +1,168 @@
+                    if mt == aiohttp.WSMsgType.TEXT:
+                        await ws_to.send_str(md)
+                    elif mt == aiohttp.WSMsgType.BINARY:
+                        await ws_to.send_bytes(md)
+                    elif mt == aiohttp.WSMsgType.PING:
+                        await ws_to.ping()
+                    elif mt == aiohttp.WSMsgType.PONG:
+                        await ws_to.pong()
+                    elif ws_to.closed:
+                        await ws_to.close(code=ws_to.close_code, message=msg.extra)
+                    else:
+                        raise ValueError('unexpected message type: %s' % msg)
+
+            # keep forwarding websocket data in both directions
+            await asyncio.wait([ws_forward(ws_server, ws_client), ws_forward(ws_client, ws_server)], return_when=asyncio.FIRST_COMPLETED)
+        return ws_server
+
+    async def proxy_handler(self, req):
+        method = getattr(self.session, req.method.lower())
+        upstream_url = self.next_url + req.path
+        headers = req.headers.copy()
+        query = req.query
+        try:
+            async with method(
+                    upstream_url,
+                    headers=headers,
+                    params=query,
+                    allow_redirects=False,
+                    # note that request.content is a StreamReader, so the data is streamed
+                    # and not fully loaded in memory (unlike with python-requests)
+                    data=req.content) as request:
+                response = aiohttp.web.StreamResponse(
+                    status=request.status, headers=request.headers)
+                writer = await response.prepare(req)
+                while True:
+                    chunk = await request.content.readany()
+                    if not chunk:
+                        break
+                    # using writer.write instead of response.write saves a few checks
+                    await writer.write(chunk)
+                return response
+        except aiohttp.ClientConnectionError as e:
+            return self.connection_error(e)
+
+    def connection_error(self, error):
+        return aiohttp.web.Response(text='Unable to connect to upstream server {} ({!s})'.format(
+            self.next_url, error), status=502)
+
+    async def index_handler(self, req):
+        if self.config is None:  # on first get, we prepare the config from upstream
+            try:
+                async with self.session.get(self.next_url) as request:
+                    index = await request.content.read()
+            except aiohttp.ClientConnectionError as e:
+                return self.connection_error(e)
+            # hack to parse the configjson from upstream buildbot config
+            start_delimiter = b'angular.module("buildbot_config", []).constant("config", '
+            start_index = index.index(start_delimiter)
+            last_index = index.index(b')</script></html>')
```

Comment:
Could be extracted to a separate function `extract_config`. Right now it takes a lot of focus to understand what the code does.